include/envoy/config/watch_map.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <set>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/subscription.h"
+
+namespace Envoy {
+namespace Config {
+
+struct Watch;
+using WatchPtr = std::unique_ptr<Watch>;
+
+struct AddedRemoved {
+  AddedRemoved(std::set<std::string> added, std::set<std::string> removed)
+      : added_(added), removed_(removed) {}
+  std::set<std::string> added_;
+  std::set<std::string> removed_;
+};
+
+class WatchMap {
+public:
+  virtual ~WatchMap() = default;
+
+  // Adds 'callbacks' to the WatchMap, with no resource names being watched.
+  // (Use updateWatchInterest() to add some names).
+  // Returns the newly added watch, to be used for updateWatchInterest. Destroy to remove from map.
+  virtual WatchPtr addWatch(SubscriptionCallbacks& callbacks) PURE;
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in added_.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in removed_.
+  virtual AddedRemoved updateWatchInterest(Watch* watch,
+                                           const std::set<std::string>& update_to_these_names) PURE;
+
+  // Intended to be called by the Watch's destructor.
+  // Expects that the watch to be removed has already had all of its resource names removed via
+  // updateWatchInterest().
+  virtual void removeWatch(Watch* watch) PURE;
```

Comment:
I would still prefer a RAII pattern here. This is a pretty clear example of resource reclamation, and fits the C++ RAII paradigm closely. This gets back to the earlier point about the parent pointer; we do this in a bunch of places in Envoy, the idea is that resources know how to clean themselves up and contain the necessary logic to do so via scope removal. This is a safer pattern (in particular in the presence of exception) and avoids the client code needing to futz with handles.

Reply:
BTW, I feel we're circling around quite a bit on this one. Maybe the best thing to do at this point is bring in another reviewer who can break deadlock, since I'm probably just being stubborn. @envoyproxy/maintainers is anyone interested in providing some additional perspective on this PR (in the space of refactoring around incremental xDS)?

Reply:
I don't understand what you're asking for. removeWatch() is only called by the destructor of Watch. Its purpose is to clean up WatchMap internal state in response to the Watch being removed. This all seems fully RAIIified already. I guess removeWatch() could become private, and Watch could become a friend?

Reply:
Yep, I think something like this makes sense. I.e. we shouldn't have to expose out on a public interface any removal.

Reply:
Cool, SGTM, done.

Reply:
You say done here, but it's still on the virtual interface :)

Reply:
Yes, looks like I missed that when I had Watch as a friend. However, now that Watch is more separated out, and "communicating" with WatchMap rather than WatchMap just using Watch structs as bits of state, isn't it necessary to be in the interface?